### PR TITLE
Fix isMethod bug

### LINKED ISSUE #1553: documentSymbols crash when editing @foo
## Input

```rb
# typed: true
class C
  @foo = T.let(3, Integer)
end
```

1. Start with this file
1. Put cursor directly after `@foo`
1. Type `d`

## Expected

No error

## Actual

```
Exception::raise(): main/lsp/DefLocSaver.cc:54 enforced condition ctx.owner.data(ctx)->isMethod() has failed: (no message provided)

Backtrace:
sorbet::Exception::enforce_handler(std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::basic_string_view<char, std::__1::char_traits<char> >, int) (in librealmain.so) (Exception.h:57)
sorbet::realmain::lsp::DefLocSaver::postTransformUnresolvedIdent(sorbet::core::Context, std::__1::unique_ptr<sorbet::ast::UnresolvedIdent, std::__1::default_delete<sorbet::ast::UnresolvedIdent> >) (in liblsp.so) (DefLocSaver.cc:54)
sorbet::ast::PostPonePostTransform_UnresolvedIdent<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context, true>::call(sorbet::core::Context, std::__1::unique_ptr<sorbet::ast::UnresolvedIdent, std::__1::default_delete<sorbet::ast::UnresolvedIdent> >, sorbet::realmain::lsp::DefLocSaver&) (in liblsp.so) (treemap.h:232)
sorbet::ast::TreeMapper<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context>::mapUnresolvedIdent(std::__1::unique_ptr<sorbet::ast::UnresolvedIdent, std::__1::default_delete<sorbet::ast::UnresolvedIdent> >, sorbet::core::Context) (in liblsp.so) (treemap.h:462)
sorbet::ast::TreeMapper<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in liblsp.so) (treemap.h:661)
sorbet::ast::TreeMapper<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context>::mapAssign(std::__1::unique_ptr<sorbet::ast::Assign, std::__1::default_delete<sorbet::ast::Assign> >, sorbet::core::Context) (in liblsp.so) (treemap.h:474)
sorbet::ast::TreeMapper<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in liblsp.so) (treemap.h:690)
sorbet::ast::TreeMapper<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context>::mapClassDef(std::__1::unique_ptr<sorbet::ast::ClassDef, std::__1::default_delete<sorbet::ast::ClassDef> >, sorbet::core::Context) (in liblsp.so) (treemap.h:281)
sorbet::ast::TreeMapper<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in liblsp.so) (treemap.h:672)
sorbet::ast::TreeMapper<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context>::mapInsSeq(std::__1::unique_ptr<sorbet::ast::InsSeq, std::__1::default_delete<sorbet::ast::InsSeq> >, sorbet::core::Context) (in liblsp.so) (treemap.h:601)
sorbet::ast::TreeMapper<sorbet::realmain::lsp::DefLocSaver, sorbet::core::Context>::mapIt(std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >, sorbet::core::Context) (in liblsp.so) (treemap.h:668)
std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> > sorbet::ast::TreeMap::apply<sorbet::core::Context, sorbet::realmain::lsp::DefLocSaver>(sorbet::core::Context, sorbet::realmain::lsp::DefLocSaver&, std::__1::unique_ptr<sorbet::ast::Expression, std::__1::default_delete<sorbet::ast::Expression> >) (in liblsp.so) (treemap.h:712)
sorbet::realmain::lsp::tryApplyDefLocSaver(sorbet::core::GlobalState const&, std::__1::vector<sorbet::ast::ParsedFile, std::__1::allocator<sorbet::ast::ParsedFile> >&) (in liblsp.so) (updates.cc:197)
sorbet::realmain::lsp::LSPLoop::runQuery(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >, sorbet::core::lsp::Query const&, std::__1::vector<sorbet::core::FileRef, std::__1::allocator<sorbet::core::FileRef> > const&) const (in liblsp.so) (updates.cc:382)
sorbet::realmain::lsp::LSPLoop::setupLSPQueryByLoc(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >, std::__1::basic_string_view<char, std::__1::char_traits<char> >, sorbet::realmain::lsp::Position const&, sorbet::realmain::lsp::LSPMethod, bool) const (in liblsp.so) (lsp.cc:57)
sorbet::realmain::lsp::LSPLoop::handleTextDocumentCompletion(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >, sorbet::realmain::lsp::MessageId const&, sorbet::realmain::lsp::CompletionParams const&) const (in liblsp.so) (completion.cc:203)
sorbet::realmain::lsp::LSPLoop::processRequestInternal(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >, sorbet::realmain::lsp::LSPMessage const&) (in liblsp.so) (request_dispatch.cc:217)
sorbet::realmain::lsp::LSPLoop::processRequest(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >, sorbet::realmain::lsp::LSPMessage const&) (in liblsp.so) (request_dispatch.cc:18)
sorbet::realmain::lsp::LSPLoop::runLSP() (in liblsp.so) (protocol.cc:205)
sorbet::realmain::realmain(int, char**) (in librealmain.so) (realmain.cc:455)
main (in sorbet) (main.cc:7)
start (in libdyld.dylib) + 1
```

### PR DESCRIPTION
**Pull request by azdavis**
This would close #1553, but I don't understand why this issue is happening in the first place.

I'd like to understand the root cause first, because this "fix" might just be papering over the actual root cause.

The first few commits are pre-work to clean up, the last one is the actual fix.

### Motivation
To fix a bug.

### Test plan
I'm not sure how to write a LSP 'edit' test (`.rbupdate`?).

### CODE REVIEW

**Comment by azdavis** [comment]
UnWIPing this, but I'm still not totally confident I understand why this change is necessary.

In builder_walk, the owner of the context when handing unresolved ivars is indeed always a method, because the ivar was either written in a method the user wrote, or put into the 'fake' Sorbet method <static-init> if the user wrote them directly on the class. Here, however, if the user wrote them directly on the class, the owner of the context is the class.

Maybe this is just the way `ast::TreeMap::apply` works and I don't need to worry about it, but I'm wondering why in this case, when unresolved ivars at the top level, the context owner is not <static-init>?

Also re: testing strategy, the test infra required to write the test I want to write (which would assert that, after applying a rbupdate, that rbupdate's document symbols are what we expect) will probably take a non-trivial amount of effort to write which I don't have time for (my internship ends soon).

**Comment by DarkDimius** [comment]
> I'd like to understand the root cause first, because this "fix" might just be papering over the actual root cause.

+1, this is papering over the bug. Let's find the actual reason.

**Comment by azdavis** [comment]
From slack and Dmitry's sleuthing: we think that `runQuery` is not running `flatten`, which is the step that would add and populate `<static-init>`.

**Comment by azdavis** [comment]
It turns out Dmitry was right, and we need to run flatten before running `tryApplyDefLocSaver`. Since flatten is run as part of `pipeline::typecheck`, just swap that to come first.

**Comment by azdavis** [comment]
hmm, maybe we actually want those to run before typechecking? tests are failing.

File: main/lsp/updates.cc [hunk 1]
```diff
@@ -378,9 +378,11 @@ LSPLoop::QueryRun LSPLoop::runQuery(unique_ptr<core::GlobalState> gs, const core
     ENFORCE(gs->lspQuery.isEmpty());
     gs->lspQuery = q;
     auto resolved = pipeline::incrementalResolve(*gs, move(updatedIndexed), opts);
+    // pipeline::typecheck must be run before tryApplyDefLocSaver, because pipeline::typecheck runs flatten, which
+    // introduces the invariant that ivars defined directly on a class are put into <static-init>.
+    pipeline::typecheck(gs, move(resolved), opts, workers);
```
Comment (jvilk-stripe) [inline main/lsp/updates.cc]: You're `move`-ing `resolved` here, and then later using `resolved`. This is an error.

I think you have to keep typechecking where it is and call flatten explicitly to transform the AST.

Comment (jvilk-stripe) [reply main/lsp/updates.cc]: Actually, would `resolved = pipeline::typecheck(...)` work?

Comment (azdavis) [reply main/lsp/updates.cc]: dang, i should've been more careful with those moves. i'll try playing around with it.

Comment (azdavis) [reply main/lsp/updates.cc]: it seems like adding `resolved = ...` fixes the errors with moving, but causes other tests to fail. we may be depending on those try* functions running before typecheck?

**Comment by azdavis** [comment]
the investigation continues: it seems that even when we run flatten separately, we now get messed-up loc information?

**Comment by azdavis** [comment]
current investigation:

given the file

```rb
# typed: true

# doc for A
class A; end
```

there are no issues, but given the file

```rb
# typed: true

# doc for A
class A; end

class B; end
```

hovering over various places (the `class` keywords, the `end` keywords) gives a hover for `class A`
hmm

File: main/lsp/updates.cc [hunk 2]
```diff
@@ -176,6 +177,13 @@ void LSPLoop::reIndexFromFileSystem() {
     }
 }
 
+void flattenAll(const core::GlobalState &gs, vector<ast::ParsedFile> &indexedCopies) {
```
Comment (jvilk-stripe) [inline main/lsp/updates.cc]: `flattenAll` mutates the ASTs, so this should consume a `vector<ast::ParsedFile>` and return a new `vector<ast::ParsedFile>`. As-is, I *think* you're moving away all of the old trees and throwing away the new ones.

**Comment by azdavis** [comment]
for the problematic file above:

before flatten:

<details>

```
ClassDef{
  kind = class
  name = EmptyTree<<C <U <root>>>>
  ancestors = []
  rhs = [
    InsSeq{
      stats = [
        Send{
          recv = ConstantLit{
            orig = nullptr
            symbol = ::Sorbet::Private::Static
          }
          fun = <U keep_for_ide>
          block = nullptr
          args = [
            ConstantLit{
              orig = UnresolvedConstantLit{
                scope = EmptyTree
                cnst = <C <U A>>
              }
              symbol = ::A
            }
          ]
        }
      ],
      expr = ClassDef{
        kind = class
        name = ConstantLit{
          orig = UnresolvedConstantLit{
            scope = EmptyTree
            cnst = <C <U A>>
          }
          symbol = ::A
        }<<C <U A>>>
        ancestors = [ConstantLit{
            orig = nullptr
            symbol = ::<todo sym>
          }]
        rhs = [
        ]
      }
    }

    InsSeq{
      stats = [
        Send{
          recv = ConstantLit{
            orig = nullptr
            symbol = ::Sorbet::Private::Static
          }
          fun = <U keep_for_ide>
          block = nullptr
          args = [
            ConstantLit{
              orig = UnresolvedConstantLit{
                scope = EmptyTree
                cnst = <C <U B>>
              }
              symbol = ::B
            }
          ]
        }
      ],
      expr = ClassDef{
        kind = class
        name = ConstantLit{
          orig = UnresolvedConstantLit{
            scope = EmptyTree
            cnst = <C <U B>>
          }
          symbol = ::B
        }<<C <U B>>>
        ancestors = [ConstantLit{
            orig = nullptr
            symbol = ::<todo sym>
          }]
        rhs = [
        ]
      }
    }
  ]
}
```

</details>

after flatten:

<details>

```
InsSeq{
  stats = [
    ClassDef{
      kind = class
      name = EmptyTree<<C <U <root>>>>
      ancestors = []
      rhs = [
        MethodDef{
          flags = self
          name = <U <static-init>><<N <U <static-init>> $111>>
          args = [Local{
              localVariable = <U <blk>>
            }]
          rhs = InsSeq{
            stats = [
              InsSeq{
                stats = [
                  Send{
                    recv = ConstantLit{
                      orig = nullptr
                      symbol = ::Sorbet::Private::Static
                    }
                    fun = <U keep_for_ide>
                    block = nullptr
                    args = [
                      ConstantLit{
                        orig = UnresolvedConstantLit{
                          scope = EmptyTree
                          cnst = <C <U A>>
                        }
                        symbol = ::A
                      }
                    ]
                  }
                ],
                expr = EmptyTree
              }
              InsSeq{
                stats = [
                  Send{
                    recv = ConstantLit{
                      orig = nullptr
                      symbol = ::Sorbet::Private::Static
                    }
                    fun = <U keep_for_ide>
                    block = nullptr
                    args = [
                      ConstantLit{
                        orig = UnresolvedConstantLit{
                          scope = EmptyTree
                          cnst = <C <U B>>
                        }
                        symbol = ::B
                      }
                    ]
                  }
                ],
                expr = EmptyTree
              }
            ],
            expr = EmptyTree
          }
        }
      ]
    }
    ClassDef{
      kind = class
      name = ConstantLit{
        orig = UnresolvedConstantLit{
          scope = EmptyTree
          cnst = <C <U A>>
        }
        symbol = ::A
      }<<C <U A>>>
      ancestors = [ConstantLit{
          orig = nullptr
          symbol = ::<todo sym>
        }]
      rhs = [
      ]
    }
    ClassDef{
      kind = class
      name = ConstantLit{
        orig = UnresolvedConstantLit{
          scope = EmptyTree
          cnst = <C <U B>>
        }
        symbol = ::B
      }<<C <U B>>>
      ancestors = [ConstantLit{
          orig = nullptr
          symbol = ::<todo sym>
        }]
      rhs = [
      ]
    }
  ],
  expr = EmptyTree
}
```

</details>

**Pull request closed without merging**